repose-aggregator/components/filters/saml-policy-translation-filter/src/main/scala/org/openrepose/filters/samlpolicy/SamlPolicyTranslationFilter.scala

```diff
@@ -21,21 +21,184 @@
 package org.openrepose.filters.samlpolicy
 
 import javax.inject.{Inject, Named}
-import javax.servlet.{FilterChain, ServletRequest, ServletResponse}
+import javax.servlet.http.HttpServletResponse.SC_INTERNAL_SERVER_ERROR
+import javax.servlet.http.{HttpServletRequest, HttpServletResponse}
+import javax.servlet.{FilterChain, ServletInputStream, ServletRequest, ServletResponse}
+import javax.ws.rs.core.MediaType
 
+import com.google.common.cache.{Cache, LoadingCache}
+import com.typesafe.scalalogging.slf4j.LazyLogging
+import net.sf.saxon.s9api.XsltExecutable
+import org.openrepose.commons.utils.http.CommonHttpHeader.CONTENT_TYPE
+import org.openrepose.commons.utils.servlet.http.{HttpServletRequestWrapper, HttpServletResponseWrapper, ResponseMode}
 import org.openrepose.core.filter.AbstractConfiguredFilter
 import org.openrepose.core.services.config.ConfigurationService
+import org.openrepose.core.services.serviceclient.akka.AkkaServiceClientFactory
 import org.openrepose.filters.samlpolicy.config.SamlPolicyConfig
+import org.openrepose.nodeservice.atomfeed.{AtomFeedListener, AtomFeedService, LifecycleEvents}
+import org.w3c.dom.Document
 
 /**
   * Created by adrian on 12/12/16.
   */
 @Named
-class SamlPolicyTranslationFilter @Inject()(configurationService: ConfigurationService)
-  extends AbstractConfiguredFilter[SamlPolicyConfig](configurationService){
+class SamlPolicyTranslationFilter @Inject()(configurationService: ConfigurationService,
+                                            atomFeedService: AtomFeedService,
+                                            akkaServiceClientFactory: AkkaServiceClientFactory)
+  extends AbstractConfiguredFilter[SamlPolicyConfig](configurationService)
+  with LazyLogging
+  with AtomFeedListener {
 
   override val DEFAULT_CONFIG: String = "saml-policy.cfg.xml"
   override val SCHEMA_LOCATION: String = "/META-INF/config/schema/saml-policy.xsd"
 
-  override def doWork(request: ServletRequest, response: ServletResponse, chain: FilterChain): Unit = chain.doFilter(request, response)
+  private var cache: LoadingCache[String, XsltExecutable] = _
+
+  override def doWork(servletRequest: ServletRequest, servletResponse: ServletResponse, chain: FilterChain): Unit = {
+    try {
+      val request = new HttpServletRequestWrapper(servletRequest.asInstanceOf[HttpServletRequest])
+      var response = servletResponse.asInstanceOf[HttpServletResponse]
+      val samlResponse = decodeResponse(request)
+      val rawDocument = readToDom(samlResponse)
+
+      val version = determineVersion(rawDocument)
+      val finalDocument = version match {
+        case 1 =>
+          request.addHeader("Identity-API-Version", "1.0")
+          rawDocument
```

Comment:
If you did go the other way, wouldn't you also have to duplicate all the header manipulation stuff? That seems like a fair amount of duplication all told. I like the way you did it, personally.

If we were going to have to support even more versions, I might consider other approaches. Namely, creating a new type with functions to process the request and response. Then you only end up doing the version check once. That doesn't really matter right now, though.